Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup/FinTwo.lean

```diff
@@ -0,0 +1,186 @@
+
+lemma sub_scalar_sq_eq_disc [NeZero (2 : K)] :
+    (m - scalar _ (m.trace / 2)) ^ 2 = scalar _ (m.disc / 4) := by
+  simp only [scalar_apply, trace_fin_two, disc_fin_two, trace_fin_two,
+    det_fin_two, sq, (by norm_num : (4 : K) = 2 * 2)]
+  ext i j
+  fin_cases i <;>
+  fin_cases j <;>
+  · simp [Matrix.mul_apply]
+    field_simp
+    ring
+
+variable (m) in
+/-- The unique eigenvalue of a parabolic matrix (junk if `m` is not parabolic). -/
+def parabolicEigenvalue : K := m.trace / 2
+
+lemma IsParabolic.sub_eigenvalue_sq_eq_zero [NeZero (2 : K)] (hm : m.IsParabolic) :
+    (m - scalar _ m.parabolicEigenvalue) ^ 2 = 0 := by
+  simp [parabolicEigenvalue, -scalar_apply, sub_scalar_sq_eq_disc, hm.2]
+
+/-- Characterization of parabolic elements: they have the form `a + m` where `a` is scalar and
+`m` is nonzero and nilpotent. -/
+lemma isParabolic_iff_exists [NeZero (2 : K)] :
+    m.IsParabolic ↔ ∃ a n, m = scalar _ a + n ∧ n ≠ 0 ∧ n ^ 2 = 0 := by
+  constructor
+  · exact fun hm ↦ ⟨_, _, (add_sub_cancel ..).symm, sub_ne_zero.mpr fun h ↦ hm.1 ⟨_, h.symm⟩,
+      hm.sub_eigenvalue_sq_eq_zero⟩
+  · rintro ⟨a, n, hm, hn0, hnsq⟩
+    constructor
+    · refine fun ⟨b, hb⟩ ↦ hn0 ?_
+      rw [← sub_eq_iff_eq_add'] at hm
+      simpa only [← hm, ← hb, ← map_sub, ← map_pow, ← map_zero (scalar (Fin 2)), scalar_inj,
+        sq_eq_zero_iff] using hnsq
+    · suffices scalar (Fin 2) (m.disc / 4) = 0 by
+        rw [← map_zero (scalar (Fin 2)), scalar_inj, div_eq_zero_iff] at this
+        have : (4 : K) ≠ 0 := by simpa [show (4 : K) = 2 ^ 2 by norm_num] using NeZero.ne _
+        tauto
+      rw [← sub_scalar_sq_eq_disc, hm, trace_add, scalar_apply, trace_diagonal]
+      simp [mul_div_cancel_left₀ _ (NeZero.ne (2 : K)),
+        (Matrix.isNilpotent_trace_of_isNilpotent ⟨2, hnsq⟩).eq_zero , hnsq]
+
+end Field
+
+section LinearOrderedRing
+
+variable {R : Type*} [CommRing R] [Nontrivial R] [Preorder R]
+  (m : Matrix (Fin 2) (Fin 2) R) (g : GL (Fin 2) R)
+
+/-- A `2 × 2` matrix is *hyperbolic* if its discriminant is strictly positive. -/
+def IsHyperbolic : Prop := 0 < m.disc
+
+/-- A `2 × 2` matrix is *elliptic* if its  discriminant is strictly negative. -/
+def IsElliptic : Prop := m.disc < 0
+
+variable {m}
+
+lemma isHyperbolic_conj_iff : (g * m * g⁻¹).IsHyperbolic ↔ m.IsHyperbolic := by
+  simp only [IsHyperbolic, disc_conj]
+
+lemma isHyperbolic_conj'_iff : (g⁻¹ * m * g).IsHyperbolic ↔ m.IsHyperbolic := by
+  simpa using isHyperbolic_conj_iff g⁻¹
+
+lemma isElliptic_conj_iff : (g * m * g⁻¹).IsElliptic ↔ m.IsElliptic := by
+  simp only [IsElliptic, disc_conj]
+
+lemma isElliptic_conj'_iff : (g⁻¹ * m * g).IsElliptic ↔ m.IsElliptic := by
+  simpa using isElliptic_conj_iff g⁻¹
+
+end LinearOrderedRing
+
+namespace GeneralLinearGroup
+
+variable {R K : Type*} [CommRing R] [Field K]
+
+/-- Synonym of `Matrix.IsParabolic`, for dot-notation. -/
+@[reducible] def IsParabolic (g : GL (Fin 2) R) : Prop := g.val.IsParabolic
```

Comment:
This is exactly the meaning of `abbrev` isn't it?
```suggestion
abbrev IsParabolic (g : GL (Fin 2) R) : Prop := g.val.IsParabolic
```
Same remark for `IsElliptic` and `IsHyperbolic` below.